File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        #store the pref and postfix multiplication first 
        preMult = [1] * (len(nums)+1)
        postMult = [1] * (len(nums)+1)
        prod = 1
        for i in range(len(nums)-1, -1, -1): 
            prod *= nums[i]
            postMult[i] *= prod
        prod = 1
        for i in range(len(nums)): 
            prod *= nums[i]
            preMult[i+1] *= prod
        
        print(preMult)
        print(postMult)
        res = [0] * len(nums)
        for i in range(len(nums)): 
            res[i] = preMult[i] * postMult[i+1]
        return res 
            

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int prefix_prod = 1; 
        int post_prod = 1; 
        vector<int> res(nums.size(), 1); 

        for(int i = 0; i < nums.size(); i++){
            res[i] *= prefix_prod; 
            prefix_prod *= nums[i]; 
        }

        for(int i = nums.size() -1; i > -1; i--){
            res[i] *= post_prod; 
            post_prod *= nums[i]; 
        }

        return res; 
    

    }
};
